architecture.md

```diff
@@ -0,0 +1,126 @@
+
+*For an example of Job + Sync, see the implementation of [`.jobs.ListUserCoursesJob`][ListUserCoursesJob]*
+
+
+## Common features
+
+### RecyclerView
+
+Implementing a `RecyclerView.Adapter` is pretty simple, and we've made it even easier! Just extend [`BaseAdapter`][BaseAdapter], and most of the boilerplate code is handled for you.
+
+*For an example, see the implementation of [`TopicListAdapter`][TopicListAdapter]*
+
+### HTML-Content
+
+For displaying HTML-based formatted content, please use [`ContentWebView`][ContentWebView]. It takes care of loading internal content (using authentication), correct opening of links (external vs internal) and resizes content designed for larger screens.
+
+External links are opened using [`openUrl`][WebUtils], which creates a `CustomTab` behind the scenes.
+
+*If the displayed content should be non-interactive, use [`PassiveWebView`][PassiveWebView] instead. It works the same but doesn't catch any click or scroll events.*
+
+
+### Share
+
+For sharing the view of an activity or fragment, just overwrite the property `url`. That's it! The base classes create a menu item handle the share action, passing the `url`-String. The current toolbar title is used as the share-`Intent`s subject.
+
+
+### Refresh
+
+To implement refreshing behaviour in an activity or fragment, simply set the property [`swipeRefreshLayout`][BaseActivity] to the `SwipeRefreshLayout` (SRL from now on) instance, and override refresh(). The base classes handle SRL styling, toggle SRL state, add a menu item for refresh and perform the actual refresh.
+
+*For an example, see the implementation of [`CourseListFragment`][CourseListFragment]*
+
+
+### Feedback to the user
+
+#### Success/Error
+
+Indicating a **successful** action to the user is done by calling [`showGenericSuccess`][DialogUtils], passing a string or string resource. This will create a toast.
+
+An **error** can be shown by calling [`showGenericError`][DialogUtils], passing a string or string resource. This will create a toast, and the error message is automagically prefixed with "Error: " for consistency.
+
+*If other ways of notifying the user of an error are available, they should generally be preferred. Examples include `@style/Content_Text.Empty` for empty content, `TextView.setError()` for inputs, etc.*
+
+#### Progress dialog
+
+Showing an indeterminate progress dialog is as simple as wrapping a `suspend`-function with [`withProgressDialog`][DialogUtils], passing a string or string resource shown in the dialog to inform the user.
+
+```kotlin
+withProgressDialog("Downloading file") {
+    val result = performDownload().await()
+    if (!result) {
+        showGenericError("Download encountered an error. Please try again.")
+        return@withProgressDialog
+    }
+    showGenericSuccess("File downloaded!")
```

Comment:
3 non-localized hardcoded strings here. Please never propose this as a good practice. I would propose to remove this example and point to the file like you did with the others.